### PR TITLE
feat(router) route by header

### PR DESCRIPTION
**Pull request by nateslo**
### Summary

This PR adds route-by-header functionality to the router. Requests can be routed by arbitrary headers in addition to the currently supported `host` header routing. `Routes.headers` field has been added to store this configuration.`Routes.headers` is a map of string header keys pointing to arrays of header values.

```

local Routes = {
  -- can contain `host` or other headers for routing
  headers = {
    [header_name_str] = { header_value_str,  header_value_str, ... }, 
  },

  -- remains for backwards compatibility
  hosts = { host, host, ... },
}
```

Existing host routing behavior remains, with the addition of being able specify hosts in either the `Route.hosts` field or as an entry of `host` in the new `Router.headers` map. `Routes.hosts` field remains for backwards compatibility. Only one of `Route.hosts` or `Router.headers.host` must be set to route-by-host. If host is set in both, route matching occurs on `hosts` configured in either. 

### Full changelog

- update router to route by req headers
- add `Routes.headers` field to `Routes` entity
- update tests
- validate wildcard hosts in `Routes.hosts` and `Routes.headers.host` with `typedefs.wildcard_host`
- update `typedefs.wildcard_host.custom_validator` to handle edge cases previously handled by inline `Routes.hosts` schema validation.
- update wildcard host validation error message from `typedefs.wildcard_host.custom_validator` for clarity.

### Issues resolved

- `typedefs.wildcard_host` is now used for checking wildcard hosts in `Routes.hosts` field as well as `Routes.headers.host`. `typedefs.wildcard_host.custom_validator` has been updated to handle [these edge cases](https://github.com/Kong/kong/pull/4758/files#diff-c09a038b3efe67503dbe9c8745616360R40). 
- error message returned from `typedefs.wildcard_host.custom_validator` when encountering multiple wildcards has been changed for clarity [here](https://github.com/Kong/kong/pull/4758/files#diff-c09a038b3efe67503dbe9c8745616360R55).

TODO: needs more tests.

### CODE REVIEW

**Comment by kikito** [comment]
I have done the following changes to this PR so far:
* I have regrouped the commits into just two (one for the functionality, one for the tests).
* I also rebased this branch over `next`, which now includes #4768 (recursive params in form-url-encoded). This allowed me to activate one of the tests which was waiting for this change.

**Comment by thibaultcha** [comment]
@kikito There is no reason for tests to be in the same commit though. On top of that, there are reasons to separate schema/migration changes to the Routes model and pure router changes. Can we revert the squashing please?

**Comment by kikito** [comment]
@thibaultcha I have done as requested. I have also included the [small changes you pushed yesterday](https://github.com/Kong/kong/compare/d3c698ad7bbb0fcabd3961b7c1d27ec86eac3c41..f371bef9465671cb3f8c83827136d387f26e34d7). The test you changed on that PR was incomplete (it did not test url-encoded content type). So I replaced that with a different version, which does test things with this PR.

I also fixed some commit messages (spacing) while I was there.

I tried to rebase over the current `next`, but I was not able to handle the new ... "surprises" we have there, so I cancelled the attempt after timeboxing it to 1 hour.

For what is worth, here's my reasoning for the changes I made to Nate's history:

* The two functional commits don't look like a clear-cut "schema changes vs code changes" split. Both of them have code which is not limited to migrations. I interpreted the second commit as a refinement to the first.
* With regards to test commits, I thought they would be more easily mergeable / rebaseable as a single unit instead of several.

**Comment by thibaultcha** [comment]
I had already properly squashed for a proper git history and pushed the commits in this PR...

Anyway, I am working on updates to this PR atm.

File: kong/router.lua [hunk 1]
```diff
@@ -969,6 +1052,15 @@ function _M.new(routes)
         return r1.submatch_weight > r2.submatch_weight
       end
 
+      do
+        local r1_n_headers = #r1.headers
+        local r2_n_headers = #r2.headers
+
+        if r1_n_headers ~= r2_n_headers then
+          return r1_n_headers > r2_n_headers
+        end
```
[the remark this answers is not in the document]
Comment (thibaultcha) [reply kong/router.lua]: No, that is not equivalent in the context of a `table.sort` comparison function; the above would produce an error.

File: spec/02-integration/05-proxy/02-router_spec.lua [hunk 2]
```diff
@@ -889,6 +889,224 @@ for _, strategy in helpers.each_strategy() do
...
+
+        local res = assert(proxy_client:send {
+          method  = "GET",
+          path    = "/",
+          headers = {
+            ["Host"]       = "domain.org",
+            ["version"]    = "v1",
+            ["kong-debug"] = 1,
+          }
+        })
```
[the remark this answers is not in the document]
Comment (thibaultcha) [reply spec/02-integration/05-proxy/02-router_spec.lua]: Yes, I am aware, but this file does not use those helpers so maintenance is easier for now (all tests look similar). We can update the suite to use the helpers if we feel like it (that said, I still prefer the original syntax for various reasons, but no strong feelings about one or the other).

**Comment by thibaultcha** [comment]
> I think that `refactor(router) accept routes tables with 'hosts' and 'headers'` should not be a separate commit, given that it refactors functionality introduced by `feat(router) route by any request header`.

This refactor has been a TODO item (explicitely stated in the code as such) since the router was first written. It was not introduced by the main feature commit of this PR. Besides, making it a separate commit simplifies the merging task for other maintainers or derived codebases (internally), and allows for an easy revert on our side if we feel like not including it. It should be separate as per our commit guidelines since it isn't required for the feature, potentially breaking (for maintainers), and needs visibility.

> I feel similarly about `tests(router) remove a duplicated test` - I would just merge it with `tests(router) add router integration tests for headers routing`.

It seems like this duplicated test is the result of a rebase gone wrong at some point in this branch. I will remove it if I can make sure that it isn't present in master/next.

> Finally, if `fix(router) ensure reducers respect matching priorities` fixes a problem in `next`, I would leave it as a single commit. On the other hand, if it fixes a problem which only exists on this PR, then I would merge it into `feat(router) route by any request header`.

It fixes a problem in next, hence why it is a separate commit.

If anybody has the final state of the git history in mind when proposing PR, I believe that would be me, having drafted those guidelines in the first place. Thanks for the review nonetheless!

**Comment by thibaultcha** [comment]
> It seems like this duplicated test is the result of a rebase gone wrong at some point in this branch. I will remove it if I can make sure that it isn't present in master/next.

I double checked, the test is also duplicated in master and next. It is correct for it to be a separate commit, as removing it as part of another commit in this PR would be incorrect.
